packages/compiler/src/core/checker.ts

```diff
@@ -5253,6 +5253,10 @@ export function createChecker(program: Program): Checker {
       return source.kind === ReflectionNameToKind[target.name];
     }
 
+    if (source.kind === "Boolean" && target.kind === "Scalar" && target.name === "boolean") {
```

Comment:
are you sure this is where the issue is, template constraint do validate that correctly already https://cadlplayground.z22.web.core.windows.net/?c=bW9kZWwgRm9vPFQgZXh0ZW5kcyBib29sZWFuPiB7CiAgZm9vOiBUCn0KCsYrVGVzdMgZVHJ1ZTrFPXRydWU%2BO8YWRmFsc8cXZsQLyBhJbnZhbGlkxhoiIj47Cn0%3D&e=%40typespec%2Fopenapi3&options=%7B%7D

Reply:
it should already be doing that in the next function call 
<img width="613" alt="image" src="https://github.com/microsoft/typespec/assets/1031227/b901b865-1272-47e2-a50d-55ae1ffb2fb4">

Reply:
This change _does_ fix the behavior.

Looks like there' s a bug in areScalarsRelated.

The first time it calls, it returns true, as expected.
![image](https://github.com/microsoft/typespec/assets/5723682/d1df2956-be8e-4cce-932b-38eda75102a9)

But when it calls it again, with seemingly the same inputs, it returns false.
![image](https://github.com/microsoft/typespec/assets/5723682/ec49422f-098b-4e1a-a686-2e8d83af51d4)

Reply:
ah, could it be a projection issue? we are comparing types in different projector?

Reply:
The only difference in the call stacks for the above cases is that the case where it unexpectedly returns false occurs during projection.

Reply:
what is the stacktrace for the 2nd call?

Reply:
my guess is that in the 2nd call we are passing a projected type and it compares the projected type from the non projected type which then fails

Reply:
Yep, looks like during projection:
![image](https://github.com/microsoft/typespec/assets/5723682/a0a87507-062d-41e8-ae0f-a21f03107ef9)

The reference equal on line 5239 evaluates to false even though the two are equivalent.